templates/model/decode/decode.go

```diff
@@ -0,0 +1,185 @@
+			KindGoldenTest: &goldentestv1alpha1.Test{},
+			KindManifest:   &manifestv1alpha1.Manifest{},
+		},
+	},
+	{
+		apiVersion: "cli.abcxyz.dev/v1beta1",
+		kinds: map[string]model.ValidatorUpgrader{
+			KindTemplate:   &specv1beta1.Spec{},
+			KindGoldenTest: &goldentestv1alpha1.Test{},
+			KindManifest:   &manifestv1alpha1.Manifest{},
+		},
+	},
+}
+
+// Decode parses the given YAML contents of r into a struct and returns it. The given filename
+// is used only for error messages. The type of struct to return is determined by the "kind" field
+// in the YAML. If the given requireKind is non-empty, then we'll also validate that the "kind"
+// of the YAML file matches requireKind, and return error if not.
+func Decode(r io.Reader, filename, requireKind string) (model.ValidatorUpgrader, string, error) {
+	buf, err := io.ReadAll(r)
+	if err != nil {
+		return nil, "", fmt.Errorf("error reading file %s: %w", filename, err)
+	}
+
+	cf := &commonFields{}
+	if err := yaml.Unmarshal(buf, cf); err != nil {
+		return nil, "", fmt.Errorf("error parsing file %s: %w", filename, err)
+	}
+
+	var apiVersion string
+	if cf.NewStyleAPIVersion.Val != "" && cf.OldStyleAPIVersion.Val != "" {
+		return nil, "", cf.OldStyleAPIVersion.Pos.Errorf("must not set both apiVersion and api_version, please use api_version only")
+	}
+	if cf.NewStyleAPIVersion.Val == "" && cf.OldStyleAPIVersion.Val == "" {
+		return nil, "", fmt.Errorf(`file %s must set the field "api_version"`, filename)
+	}
+	if cf.NewStyleAPIVersion.Val != "" {
+		apiVersion = cf.NewStyleAPIVersion.Val
+	}
+	if cf.OldStyleAPIVersion.Val != "" {
+		apiVersion = cf.OldStyleAPIVersion.Val
+	}
+
+	if cf.Kind == "" {
+		return nil, "", fmt.Errorf(`file %s must set the field "kind"`, filename)
+	}
+	if requireKind != "" && cf.Kind != requireKind {
+		return nil, "", fmt.Errorf("file %s has kind %q, but %q is required", filename, cf.Kind, requireKind)
+	}
+
+	vu, err := modelForKind(filename, apiVersion, cf.Kind)
+	if err != nil {
+		return nil, "", err
+	}
+
+	if err := yaml.Unmarshal(buf, vu); err != nil {
+		return nil, "", fmt.Errorf("error parsing YAML file %s: %w", filename, err)
+	}
+
+	return vu, apiVersion, nil
+}
+
+// DecodeValidateUpgrade  parses the given YAML contents of r into a struct,
+// then repeatedly calls Upgrade() and Validate() on it until it's the newest version, then
+// returns it. requireKind has the same meaning as in Decode().
+func DecodeValidateUpgrade(r io.Reader, filename, requireKind string) (model.ValidatorUpgrader, error) {
+	vu, apiVersion, err := Decode(r, filename, requireKind)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := vu.Validate(); err != nil {
+		return nil, fmt.Errorf("validation failed in %s: %w", filename, err)
+	}
+
+	for {
+		upgraded, err := vu.Upgrade()
+		if err != nil {
+			return nil, fmt.Errorf("internal error: YAML model couldn't be upgraded from api_version %s: %w", apiVersion, err)
+		}
+		if upgraded == nil {
```

Comment:
I worry about the sentinel value "nil". Would it be better to have a specific error type that we can check instead? This is similar to how io returns `io.EOF` when it reaches the end of the file.

Reply:
Good idea, implemented.